File: src/config/reading.ts
/**
 * Adapter IPC phía webview cho Chế độ đọc — Story 5.11, FR11.
 *
 * Cùng khuôn `./chapter.ts`: một `invoke`, một `try/catch`, hình dạng BA TRẠNG THÁI
 * `{ chapter: ReadingChapter | null, error: IpcError | null }` — KHÔNG BAO GIỜ ném. Tầng
 * UI (`src/modes/readingState.ts`) hiển thị lỗi bằng `tError()`, không bằng `try/catch`.
 *
 * ⚠️ `invoke()` gửi tham số dạng camelCase; hàm này không nhận tham số nào nên điều đó
 * không áp dụng ở đây — cùng ghi chú mà `./chapter.ts::readOpenChapter` để lại.
 */
import { invoke } from '@tauri-apps/api/core'
import type { IpcError } from '../i18n'

/**
 * Một câu, cho Chế độ đọc — khớp `commands::segment::ReadingSegment` phía Rust,
 * `snake_case` trên dây.
 *
 * ⚠️ **KHÔNG** `is_omitted`: câu đã cắt bỏ không bao giờ đi qua dây tới đây — chốt lọc
 * sống ở Rust (`core::segment::omit`/`core::segment::reading`, AD-1). Một component đọc
 * kiểu này **không có** đường nào để tự lọc lần hai, vì không có gì để lọc.
 */
export type ReadingSegment = {
  id: number
  source_text: string
  target_text: string
}

/**
 * Một đoạn — khớp `commands::segment::ReadingParagraph`. Không đoạn nào rỗng: nếu mọi câu
 * của một đoạn đều đã bị cắt bỏ, đoạn đó không xuất hiện trong `paragraphs` (Rust quyết,
 * xem doc-comment của `paragraphs_in_translation`).
 */
export type ReadingParagraph = {
  segments: ReadingSegment[]
}

/**
 * Chương đang mở, đã gom đoạn — khớp `commands::segment::ReadingChapter`, `snake_case`
 * trên dây.
 */
export type ReadingChapter = {
  chapter_id: number
  chapter_ord: number
  /** `null` ⇒ Chương chưa đặt tên — cùng ngữ nghĩa `ChapterRow::title` (Story 5.7). */
  chapter_title: string | null
  paragraphs: ReadingParagraph[]
}

/** Ba trạng thái cho [`readReadingChapter`], cùng khuôn `ReadOpenChapterResult`. */
export type ReadReadingChapterResult = {
  chapter: ReadingChapter | null
  error: IpcError | null
}

function isIpcError(value: unknown): value is IpcError {
  if (typeof value !== 'object' || value === null) return false
  const v = value as Partial<IpcError>
  return (
    typeof v.code === 'string' &&
    typeof v.message_key === 'string' &&
    typeof v.retryable === 'boolean' &&
    typeof v.params === 'object' &&
    // ⚠️ Cùng canh gác mà mọi adapter khác của kho này giữ — `IpcError` là một lời khai về
    //    dữ liệu đã qua dây, không một bảo đảm của trình biên dịch.
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- xem chú thích ngay trên
    v.params !== null
  )
}

function isReadingSegment(value: unknown): value is ReadingSegment {
  if (typeof value !== 'object' || value === null) return false
  const v = value as Partial<ReadingSegment>
  return typeof v.id === 'number' && typeof v.source_text === 'string' && typeof v.target_text === 'string'
}

/** 🔴 Kiểm MỌI phần tử, không chỉ phần tử đầu — cùng lý lẽ `config/library.ts::isSearchHitArray`. */
function isReadingSegmentArray(value: unknown): value is ReadingSegment[] {
  return Array.isArray(value) && value.every(isReadingSegment)
}

function isReadingParagraph(value: unknown): value is ReadingParagraph {
  if (typeof value !== 'object' || value === null) return false
  const v = value as Partial<ReadingParagraph>
  return isReadingSegmentArray(v.segments)
}

function isReadingParagraphArray(value: unknown): value is ReadingParagraph[] {
  return Array.isArray(value) && value.every(isReadingParagraph)
}

/** Vị từ kiểm kiểu LÚC CHẠY cho toàn bộ `ReadingChapter` — kiểm MỌI trường, MỌI phần tử. */
function isReadingChapter(value: unknown): value is ReadingChapter {
  if (typeof value !== 'object' || value === null) return false
  const v = value as Partial<ReadingChapter>
  return (
    typeof v.chapter_id === 'number' &&
    typeof v.chapter_ord === 'number' &&
    (typeof v.chapter_title === 'string' || v.chapter_title === null) &&
    isReadingParagraphArray(v.paragraphs)
  )
}

/** Có cầu IPC thật hay chỉ là phiên `npm run dev` trong trình duyệt thường. */
function hasIpcBridge(): boolean {
  return typeof window !== 'undefined' && '__TAURI_INTERNALS__' in window
}

/** Lỗi hồi phòng khi Rust trượt bằng một thứ không phải `IpcError`. */
const UNKNOWN_IPC_ERROR: IpcError = {
  code: 'ipc.unknown',
  message_key: 'err.unknown',
  params: {},
  retryable: false,
}

/** Tên command trên dây. Khớp `src-tauri/src/commands/segment.rs` (module `wire`). */
const CMD_READ_READING_CHAPTER = 'read_reading_chapter'

/**
 * Đọc Chương đang mở, đã gom đoạn cho Chế độ đọc. Không ném — cùng lý do và cùng khuôn
 * [`./chapter.ts::readOpenChapter`].
 */
export async function readReadingChapter(): Promise<ReadReadingChapterResult> {
  try {
    const raw = await invoke<unknown>(CMD_READ_READING_CHAPTER)
    if (!isReadingChapter(raw)) {
      console.error(`[reading] \`${CMD_READ_READING_CHAPTER}\` trả một hình dạng không phải ReadingChapter: ${String(raw)}`)
      return { chapter: null, error: UNKNOWN_IPC_ERROR }
    }
    return { chapter: raw, error: null }
  } catch (err) {
    if (isIpcError(err)) return { chapter: null, error: err }

    // 🔴 CÓ cầu IPC mà vẫn trượt bằng một thứ không phải `IpcError` ⇒ lỗi THẬT, KHÔNG
    // phải "chạy ngoài Tauri" — cùng bài học đã ghi ở `./chapter.ts::readOpenChapter`.
    if (hasIpcBridge()) {
      console.error(`[reading] \`${CMD_READ_READING_CHAPTER}\` trượt bằng một lỗi không phải IpcError: ${String(err)}`)
      return { chapter: null, error: UNKNOWN_IPC_ERROR }
    }

    // Không có cầu IPC — `npm run dev` trong một trình duyệt thường. Không phải một lỗi
    // để hiện lên; trạng thái thứ BA của I/O Matrix.
    console.info(`[reading] không gọi được \`${CMD_READ_READING_CHAPTER}\` — chạy ngoài Tauri? ${String(err)}`)
    return { chapter: null, error: null }
  }
}

File: src-tauri/src/core/segment/reading.rs
//! Gom segment **thuộc bản dịch** thành đoạn cho Chế độ đọc — Story 5.11, FR11.
//!
//! ─────────────────────────────────────────────────────────────────────────────
//! VÌ SAO HÀM NÀY Ở RUST, KHÔNG MỘT `computed` Ở VUE — cùng kỷ luật `paragraph.rs`
//! ─────────────────────────────────────────────────────────────────────────────
//! Cắt đoạn *trông như* việc render: duyệt danh sách, gặp `is_target_paragraph_end` thì
//! xuống đoạn. Ca biên phá cách đọc đó — đo được, không suy:
//!
//! ```text
//! câu 1  is_omitted=0  is_target_paragraph_end=0
//! câu 2  is_omitted=1  is_target_paragraph_end=1   ← người dùng đã CẮT BỎ đúng câu kết đoạn
//! câu 3  is_omitted=0  is_target_paragraph_end=0
//! ```
//!
//! Lọc trước rồi cắt sau ⇒ `[1, 3]` **một đoạn** — hai đoạn của người dùng nhập làm một,
//! im lặng. Cắt trước rồi lọc sau ⇒ đúng hai đoạn nhưng đoạn đầu phải mang cờ của một câu
//! **không còn tồn tại**. Quy tắc thật là *"cờ kết đoạn chuyển cho câu còn sống liền trước
//! trong cùng đoạn"*, và một quy tắc có ca biên thật thì không được sống ở webview (AD-1)
//! — đúng lý lẽ mà `core/segment/paragraph.rs` đã viết cho ba ca biên của AD-37.
//!
//! ─────────────────────────────────────────────────────────────────────────────
//! 🔴 CHỐT LỌC LÀ [`super::omit::segments_in_translation`], KHÔNG MỘT VỊ TỪ THỨ HAI
//! ─────────────────────────────────────────────────────────────────────────────
//! Doc-comment của `omit.rs` cấm thẳng một `v-if`/vị từ `!is_omitted` rải ở nơi khác:
//! *"chốt lọc phải có đúng một vị từ"*. Hàm dưới đây gọi lại chính chốt đó rồi ĐỐI CHIẾU
//! kết quả với dãy gốc theo con trỏ phần tử (`std::ptr::eq`) — không viết lại `!s.is_omitted`
//! một lần thứ hai ở tệp này.

use crate::commands::segment::ChapterSegment;

use super::omit;

/// Gom `segments` thành các đoạn **thuộc bản dịch** — Story 5.11.
///
/// Mỗi đoạn trả về là một dãy `&ChapterSegment` liên tiếp theo `(ord, id)`. Không đoạn
/// nào rỗng lọt ra ngoài: một đoạn mà mọi câu của nó đều bị cắt bỏ đơn giản không xuất
/// hiện trong kết quả, cùng khuôn *"không bịa"* của [`super::paragraph::merged`].
///
/// ⚠️ Nhận `&[ChapterSegment]`, trả `Vec<Vec<&ChapterSegment>>` — không nhân bản
/// `source_text`/`target_text`, cùng lý do hiệu năng mà [`omit::segments_in_translation`]
/// đã ghi cho Chương lớn nhất có thật (9.850 câu).
///
/// # Thuật toán
/// Duyệt `segments` theo đúng thứ tự đã lưu (một lượt, không chia hai lượt lọc-rồi-cắt):
/// - Câu **còn sống** (có mặt trong [`omit::segments_in_translation`]) được đẩy vào đoạn
///   đang gom; cờ `is_target_paragraph_end` của CHÍNH câu đó đóng đoạn lại.
/// - Câu **đã cắt bỏ** không được đẩy vào đâu cả — nó vắng mặt hoàn toàn. Nhưng nếu nó
///   mang cờ kết đoạn VÀ đoạn đang gom không rỗng, cờ ấy đóng đoạn đang gom ngay tại đó:
///   đây chính là phép *"chuyển cờ cho câu còn sống liền trước"*. Đoạn đang gom rỗng
///   (mọi câu trước đó trong đoạn cũng đã bị cắt bỏ, hoặc đoạn trước vừa đóng xong) thì
///   không có gì để đóng — đúng ca *"cả một đoạn bị cắt bỏ"*.
#[must_use]
pub fn paragraphs_in_translation(segments: &[ChapterSegment]) -> Vec<Vec<&ChapterSegment>> {
    // 🔴 Chốt lọc DUY NHẤT — xem doc-comment đầu tệp. `surviving` giữ THAM CHIẾU vào chính
    // các phần tử của `segments`, nên `std::ptr::eq` dưới đây so được từng phần tử một mà
    // không cần một vị từ `!is_omitted` thứ hai.
    let surviving = omit::segments_in_translation(segments);

    let mut result: Vec<Vec<&ChapterSegment>> = Vec::new();
    let mut current: Vec<&ChapterSegment> = Vec::new();
    let mut next_surviving = 0usize;

    for segment in segments {
        let is_surviving =
            next_surviving < surviving.len() && std::ptr::eq(surviving[next_surviving], segment);

        if is_surviving {
            next_surviving += 1;
            current.push(segment);
            if segment.is_target_paragraph_end {
                result.push(std::mem::take(&mut current));
            }
            continue;
        }

        // Câu đã CẮT BỎ — không đẩy vào đâu cả (AC5 · FR133: vắng mặt hoàn toàn, không
        // chỗ trống, không `[…]`). Cờ kết đoạn của nó, nếu có, chuyển cho câu còn sống
        // liền trước trong `current` bằng cách đóng đoạn ĐANG GOM ngay tại đây.
        if segment.is_target_paragraph_end && !current.is_empty() {
            result.push(std::mem::take(&mut current));
        }
    }

    // Đoạn cuối Chương không mang cờ kết đoạn (AD-37, ca biên "segment cuối Chương ⇒ tắt,
    // luôn luôn") nên nó không tự đóng trong vòng lặp — đóng nó ở đây nếu còn sót lại.
    if !current.is_empty() {
        result.push(current);
    }

    result
}
